apps/service_providers/llm_service/history_managers.py

```diff
@@ -0,0 +1,241 @@
+from apps.chat.models import ChatMessage, ChatMessageType
+from apps.experiments.models import Experiment, ExperimentSession
+from apps.pipelines.models import PipelineChatHistory, PipelineChatHistoryTypes
+
+
+class BaseHistoryManager(metaclass=ABCMeta):
+    @abstractmethod
+    def get_chat_history(self, input_messages: list):
+        pass
+
+    @abstractmethod
+    def add_messages_to_history(
+        self,
+        input: str,
+        save_input_to_history: bool,
+        input_message_metadata: dict,
+        output: str,
+        save_output_to_history: bool,
+        experiment_tag: str,
+        output_message_metadata: dict,
+    ):
+        pass
+
+    def save_message_to_history(
+        self,
+        message: str,
+        type_: ChatMessageType,
+        message_metadata: dict | None = None,
+        experiment_tag: str | None = None,
+    ):
+        """
+        Create a chat message and appends the file ids from each resource to the `openai_file_ids` array in the
+        chat message metadata.
+        Example resource_file_mapping: {"resource1": ["file1", "file2"], "resource2": ["file3", "file4"]}
+        """
+        metadata = self.get_trace_metadata()
+        metadata = metadata | (message_metadata or {})
+        chat_message = ChatMessage.objects.create(
+            chat=self.session.chat,
+            message_type=type_.value,
+            content=message,
+            metadata=metadata,
+        )
+
+        if experiment_tag:
+            tag, _ = Tag.objects.get_or_create(
+                name=experiment_tag,
+                team=self.session.team,
+                is_system_tag=True,
+                category=TagCategories.BOT_RESPONSE,
+            )
+            chat_message.add_tag(tag, team=self.session.team, added_by=None)
+
+        if type_ == ChatMessageType.AI:
+            self.ai_message = chat_message
+            chat_message.add_version_tag(
+                version_number=self.experiment_version_number, is_a_version=self.experiment_is_a_version
+            )
+
+        return chat_message
+
+    def get_trace_metadata(self) -> dict:
+        if self.trace_service:
+            trace_info = self.trace_service.get_current_trace_info()
+            if trace_info:
+                return {
+                    "trace_info": {**trace_info.model_dump(), "trace_provider": self.trace_service.type},
+                }
+        return {}
+
+
+class ExperimentHistoryManager(BaseHistoryManager):
+    def __init__(
+        self,
+        session: ExperimentSession,
+        experiment: Experiment | None = None,
+        max_token_limit: int | None = None,
+        chat_model: BaseChatModel | None = None,
+        trace_service=None,
+    ):
+        self.session = session
+        self.max_token_limit = max_token_limit
+        self.chat_model = chat_model
+        self.trace_service = trace_service
+
+        # TODO: Think about passing this in as context metadata rather
```

Comment:
We can do away with the experiment parameter here if we pass in the output message metadata from the `TopicBot` or pipeline node. No need to pass around the processor experiment just to generate tags from it. I'm keen to do this in a followup PR. I noticed that the version tags in a router setup is a bit misleading, since it will tag the output message with the processor experiment version, not the router experiment version.

To solve this, we can generate the output message tags outside of the runnable/history manager and tag output messages like this for instance:
1. the router experiment version i.e. v2 or v2-unreleased
2. the processor bot + its version. Say I have an "english" child bot, the tag will be "english-v1"

Reply:
@coderabbitai create an issue for this and assign it to me